src/librustc_mir/build/block.rs

```diff
@@ -83,4 +82,117 @@ impl<'a,'tcx> Builder<'a,'tcx> {
             block.unit()
         })
     }
+
+    pub fn stmt_expr(&mut self, mut block: BasicBlock, expr: Expr<'tcx>) -> BlockAnd<()> {
+        let this = self;
+        let expr_span = expr.span;
+        let scope_id = this.innermost_scope_id();
+        // Handle a number of expressions that don't need a destination at all. This
+        // avoids needing a mountain of temporary `()` variables.
+        match expr.kind {
+            ExprKind::Scope { extent, value } => {
+                let value = this.hir.mirror(value);
+                this.in_scope(extent, block, |this, _| this.stmt_expr(block, value))
+            }
+            ExprKind::Assign { lhs, rhs } => {
+                let lhs = this.hir.mirror(lhs);
+                let scope_id = this.innermost_scope_id();
+                let lhs_span = lhs.span;
+                let lhs_ty = lhs.ty;
+
+                let lhs_needs_drop = this.hir.needs_drop(lhs_ty);
+
+                // Note: we evaluate assignments right-to-left. This
+                // is better for borrowck interaction with overloaded
+                // operators like x[j] = x[i].
+
+                // Generate better code for things that don't need to be
+                // dropped. We need the temporary as_operand generates
+                // so we can clean up the data if evaluating the LHS unwinds,
+                // but if the LHS (and therefore the RHS) doesn't need
+                // unwinding, we just translate directly to an rvalue instead.
+                let rhs = if lhs_needs_drop {
```

Comment:
Note that `as_rvalue` will generate a temporary in some cases anyway, so this just better handles stuff like `a = b + c` where everything is an integer, previous the `b + c` would get it's own temporary.

That said, checking for both `lhs_needs_drop` and `rhs_needs_drop` isn't difficult and is certainly more robust, especially since I'm almost certain that the temporary can only be elided in that case.